### PR TITLE
[XCOFF] Use RLDs to print branches even without -r

### PR DESCRIPTION
**Pull request by stephenpeckham**
This presents misleading and confusing output.  If you have a function defined at the beginning of an XCOFF object file, and you have a function call to an external function, the function call disassembles as a branch to the local function.  That is, 

`void f() { f(); g();}`

disassembles as 
>00000000 <.f>:
       0: 7c 08 02 a6   mflr 0
       4: 94 21 ff c0   stwu 1, -64(1)                                             
       8: 90 01 00 48   stw 0, 72(1)
      c: 4b ff ff f5   bl 0x0 <.f>
      10: 4b ff ff f1   bl 0x0 <.f> 

With this PR, the second call will display:

`10: 4b ff ff f1   bl 0x0 <.g>  `

Using -r can help, but you still get the confusing output:

>10: 4b ff ff f1   bl 0x0 <.f>
      00000010:  R_RBR        .g

### CODE REVIEW

File: llvm/tools/llvm-objdump/llvm-objdump.cpp [hunk 1]
```diff
@@ -2160,19 +2198,33 @@ disassembleObject(ObjectFile &Obj, const ObjectFile &DbgObj,
...
+                if (Obj.isXCOFF() && findRel()) {
+                  SmallString<32> Val;
+
+                  // If we have a valid relocation, try to print the
+                  // corresponding symbol name. Multiple relocations on the
+                  // same instruction are not handled.
+                  if (!getRelocationValueString(*RelCur, false, Val))
+                    *TargetOS << Val;
+                  else
+                    *TargetOS << TargetName;
```
Comment (EsmeYi) [inline llvm/tools/llvm-objdump/llvm-objdump.cpp:2224]: It's better to use consumeError or reportWarning when an error is returned by getRelocationValueString().

File: llvm/tools/llvm-objdump/llvm-objdump.cpp [hunk 2]
```diff
@@ -2013,6 +2023,29 @@ disassembleObject(ObjectFile &Obj, const ObjectFile &DbgObj,
           }
         }
 
+        auto findRel = [&]() {
+          // Hexagon handles relocs in pretty printer
+          if (Obj.getArch() == Triple::hexagon)
+            return false;
```
Comment (EsmeYi) [inline llvm/tools/llvm-objdump/llvm-objdump.cpp:2029]: There is no need to put the check in findRel().

File: llvm/tools/llvm-objdump/llvm-objdump.cpp [hunk 3]
```diff
@@ -2209,36 +2261,19 @@ disassembleObject(ObjectFile &Obj, const ObjectFile &DbgObj,
...
-
-            // Stop when RelCur's offset is past the disassembled
-            // instruction/data. Note that it's possible the disassembled data
-            // is not the complete data: we might see the relocation printed in
-            // the middle of the data, but this matches the binutils objdump
-            // output.
-            if (Offset >= Index + Size)
-              break;
-
+        if (InlineRelocs) {
```
Comment (EsmeYi) [inline llvm/tools/llvm-objdump/llvm-objdump.cpp:2264]: Leave the check `Obj.getArch() != Triple::hexagon` here.

**Comment by stephenpeckham** [comment]
@EsmeYi Did you have any other comments?

File: llvm/test/tools/llvm-objdump/XCOFF/disassemble-symbolize-operands2.ll [hunk 4]
```diff
@@ -0,0 +1,62 @@
+; RUN: llc -mtriple=powerpc-ibm-aix-xcoff %s -filetype=obj -o %t
+; RUN: llvm-objdump %t -d --symbolize-operands --no-show-raw-insn \
```
Comment (EsmeYi) [inline llvm/test/tools/llvm-objdump/XCOFF/disassemble-symbolize-operands2.ll:2]: add the `-r` option?

**Comment by EsmeYi** [comment]
The improved behavior in this patch looks reasonable. However, I get some concern on the fact that the system `objdump` also prints the local function as the original behavior in `llvm-objdump`.

**Comment by stephenpeckham** [comment]
There is no system `objdump`.  The `objdump` command is a GNU tool.  Maybe someone will improve that command someday.

File: llvm/test/tools/llvm-objdump/XCOFF/disassemble-symbolize-operands2.ll (hunk 4 above)
Comment (stephenpeckham) [reply llvm/test/tools/llvm-objdump/XCOFF/disassemble-symbolize-operands2.ll]: Added -r option. This shows that the "bl" instruction and the RLD display the same function.

File: llvm/test/CodeGen/PowerPC/aix-xcoff-reloc.ll [hunk 5]
```diff
@@ -456,7 +456,7 @@ declare i32 @bar(i32)
 ; SYM-NEXT: ]
 
 
-; DIS:      {{.*}}aix-xcoff-reloc.ll.tmp.o:   file format aixcoff-rs6000
+; DIS:      :   file format aixcoff-rs6000
```
[the remark this answers is not in the document]
Comment (stephenpeckham) [reply llvm/test/CodeGen/PowerPC/aix-xcoff-reloc.ll:459]: When the temporary object file name is encoded in the FileCheck patterns, you're prevented from running commands by hand during development.  For example, I frequently create an object file first, then run

` llvm-objdump ... | FileCheck ...`

 to see if the test passes.

**Comment by EsmeYi** [comment]
LGTM, thanks.
Please wait until digger is happy with it.

**Comment by stephenpeckham** [comment]
The current llvm-objdump code does not handle fixed-up relocations. I was going to add this in a separate PR, but it really needs to be include in this PR.  A testcase requires a linked program, which is not supported yet by yaml, so a test will have to be added later. With the latest update, a branch to fixup code prints a comment for the target symbol.  The comment is not printed if -r is used.

File: llvm/test/CodeGen/PowerPC/aix-xcoff-reloc.ll (hunk 5 above)
[the remark this answers is not in the document]
Comment (diggerlin) [reply llvm/test/CodeGen/PowerPC/aix-xcoff-reloc.ll:459]: you can change to DIS: {{.*}} : file format aixcoff-rs6000.

**Review by diggerlin — APPROVED** [review]
I put two comments on the patch, I do not have further comment on the patch.

File: llvm/tools/llvm-objdump/llvm-objdump.cpp [hunk 6]
```diff
@@ -2013,6 +2022,26 @@ disassembleObject(ObjectFile &Obj, const ObjectFile &DbgObj,
...
+              ++RelCur;
+              continue;
+            }
+
+            // Stop when RelCur's offset is past the disassembled
+            // instruction/data.
+            if (RelOffset >= Index + Size)
+              return false;
+            if (RelOffset >= Index)
+              return true;
```
[the remark this answers is not in the document]
Comment (stephenpeckham) [reply llvm/tools/llvm-objdump/llvm-objdump.cpp:2039]: I could make this change, but I think the code is easier to read as it is.

[the remark this answers is not in the document]
Comment (diggerlin) [reply llvm/tools/llvm-objdump/llvm-objdump.cpp:2039]: it is suggestion , feel free to keep your code. thanks

**Pull request merged**